src/algebra/module_regular.lean

```diff
@@ -0,0 +1,184 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.group
+import algebra.group_power.basic
+import algebra.module.basic
+import algebra.smul_with_zero
+/-!
+# Action of regular elements on a module
+
+We introduce `M`-regular elements, in the context of an `R`-module `M`.
+
+By definition, an `M`-regular element of a commutative ring acting on an `R`-module `M` is an
+element `a ∈ R` such that the function `m ↦ a • m` is injective.
+-/
+
+--should it be in the module namespace?
+namespace module
+
+variables {R S : Type*} (M : Type*) {a b : R} {s : S}
+
+section has_scalar
+
+variables [has_scalar R M] [has_scalar R S] [has_scalar S M] [is_scalar_tower R S M]
+
+/-- An `M`-regular element is an element `c` such that multiplication on the left by `c` is an
+injective map `M → M`. -/
+def is_regular (c : R) := function.injective ((•) c : M → M)
+
+/-- In a monoid, the product of `M`-regular elements is `M`-regular. -/
+lemma is_regular.mul (ra : is_regular M a) (rs : is_regular M s) :
+  is_regular M (a • s) :=
+λ a b ab, rs (ra ((smul_assoc _ _ _).symm.trans (ab.trans (smul_assoc _ _ _))))
+
+/--  If an element `b` becomes `M`-regular after multiplying it on the left by an `M`-regular
+element, then `b` is `M`-regular. -/
+lemma is_regular.of_mul (ab : is_regular M (a • s)) :
+  is_regular M s :=
+@function.injective.of_comp _ _ _ (λ m : M, a • m) _ (λ c d cd, ab
+  (by rwa [smul_assoc, smul_assoc]))
+
+/--  An element is `M`-regular if and only if multiplying it on the left by an `M`-regular element
+is `M`-regular. -/
+@[simp] lemma mul_is_regular_iff (b : S) (ha : is_regular M a) :
+  is_regular M (a • b) ↔ is_regular M b :=
+⟨λ ab, is_regular.of_mul M ab, λ ab, is_regular.mul M ha ab⟩
```

Comment:
The names should reflect these are about `smul` not `mul`:
```suggestion
/-- In a monoid, the product of `M`-regular elements is `M`-regular. -/
lemma is_regular.smul (ra : is_regular M a) (rs : is_regular M s) :
  is_regular M (a • s) :=
λ a b ab, rs (ra ((smul_assoc _ _ _).symm.trans (ab.trans (smul_assoc _ _ _))))

/--  If an element `b` becomes `M`-regular after multiplying it on the left by an `M`-regular
element, then `b` is `M`-regular. -/
lemma is_regular.of_smul (ab : is_regular M (a • s)) :
  is_regular M s :=
@function.injective.of_comp _ _ _ (λ m : M, a • m) _ (λ c d cd, ab
  (by rwa [smul_assoc, smul_assoc]))

/--  An element is `M`-regular if and only if multiplying it on the left by an `M`-regular element
is `M`-regular. -/
@[simp] lemma smul_is_regular_iff (b : S) (ha : is_regular M a) :
  is_regular M (a • b) ↔ is_regular M b :=
⟨λ ab, is_regular.of_smul M ab, λ ab, is_regular.smul M ha ab⟩
```

Reply:
Ah, thanks for picking this up already!  This was a test run to see how the new file worked, but I am *very* happy that you are supervising this PR!